File: shared/java/top/fpsmaster/modules/music/IngameOverlay.java
package top.fpsmaster.modules.music;

import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.util.ResourceLocation;
import top.fpsmaster.FPSMaster;
import top.fpsmaster.features.impl.interfaces.MusicOverlay;
import top.fpsmaster.font.impl.UFontRenderer;
import top.fpsmaster.modules.music.MusicPlayer;
import top.fpsmaster.modules.music.netease.Music;
import top.fpsmaster.utils.Utility;
import top.fpsmaster.utils.math.animation.AnimationUtils;
import top.fpsmaster.utils.render.Render2DUtils;

import java.awt.Color;

public class IngameOverlay {
    private static float songProgress = 0f;
    private static double[] smoothCurve = new double[0];

    public static void onRender() {
        if (MusicPlayer.playList.getCurrent() != -1) {
            ScaledResolution sr = new ScaledResolution(Utility.mc);
            double[] curve = MusicPlayer.getCurve();
            if (curve.length != 0) {
                int numBars = 60;           // 你希望显示的频谱柱条数
                if (smoothCurve.length != numBars) {
                    smoothCurve = new double[numBars];
                }

                float width = (float) sr.getScaledWidth() / numBars;

                float screenWidth = sr.getScaledWidth();
                float screenHeight = sr.getScaledHeight();

                int binsPerBar = curve.length / numBars;

                for (int bar = 0; bar < numBars; bar++) {
                    double sum = 0.0;
                    for (int j = 0; j < binsPerBar; j++) {
                        int idx = bar * binsPerBar + j;
                        sum += Math.max(curve[idx], 0.0);
                    }
                    double averageMagnitude = sum / binsPerBar;
                    averageMagnitude = Math.min(averageMagnitude, 1.0);
                    averageMagnitude = Math.sqrt(averageMagnitude);
                    smoothCurve[bar] = AnimationUtils.base(smoothCurve[bar], averageMagnitude, 0.1);
                    float xPos = (float) bar / numBars * screenWidth;
                    float height = (float) (smoothCurve[bar] * 100f * MusicOverlay.amplitude.value.floatValue());

                    Render2DUtils.drawRect(
                            xPos,
                            screenHeight - height,
                            width,
                            height,
                            MusicOverlay.color.getColor()
                    );
                }
            }
        }
    }

    public static void drawSong(float x, float y, float width, float height) {
        ScaledResolution sr = new ScaledResolution(Utility.mc);
        Music current = (Music) MusicPlayer.playList.musics.get(MusicPlayer.playList.getCurrent());
        UFontRenderer s18 = FPSMaster.fontManager.s18;

        Render2DUtils.drawOptimizedRoundedRect(x, y, width, height, new Color(0, 0, 0, 180));
        Render2DUtils.drawOptimizedRoundedRect(x, y, songProgress, height, MusicOverlay.progressColor.getColor());

        songProgress = (float) AnimationUtils.base((double) songProgress, 6 + (width - 6) * MusicPlayer.curPlayProgress, 0.1);

        Render2DUtils.drawImage(
                new ResourceLocation("music/netease/" + current.id),
                x + 5,
                y + 5,
                height - 10,
                height - 10,
                -1
        );

        FPSMaster.fontManager.s18.drawString(
                current.name,
                x + 40,
                y + 6,
                FPSMaster.theme.getTextColorTitle().getRGB()
        );

        FPSMaster.fontManager.s16.drawString(
                current.author,
                x + 40,
                y + 18,
                FPSMaster.theme.getTextColorDescription().getRGB()
        );
    }
}

File: shared/java/top/fpsmaster/modules/music/IngameOverlay.kt
package top.fpsmaster.modules.music

import net.minecraft.client.gui.ScaledResolution
import net.minecraft.util.ResourceLocation
import top.fpsmaster.FPSMaster
import top.fpsmaster.features.impl.interfaces.MusicOverlay
import top.fpsmaster.modules.music.MusicPlayer.playProgress
import top.fpsmaster.modules.music.netease.Music
import top.fpsmaster.utils.Utility
import top.fpsmaster.utils.math.animation.AnimationUtils.base
import top.fpsmaster.utils.render.Render2DUtils
import java.awt.Color

object IngameOverlay {
    private var songProgress = 0f
    var smoothCurve: DoubleArray? = DoubleArray(0)

    fun onRender() {
        if (MusicPlayer.playList.getCurrent() != null) {
            val sr = ScaledResolution(Utility.Companion.mc)
            if (!MusicPlayer.getCurve()!!.isEmpty()) {
                val width: Float = ((sr.scaledWidth / 2f - 100) / MusicPlayer.getCurve()!!.size).coerceAtLeast(1f)
                var x = 0f
                if (smoothCurve!!.size != MusicPlayer.getCurve()!!.size) {
                    smoothCurve = MusicPlayer.getCurve()!!
                }
                for (i in 0 until MusicPlayer.getCurve()!!.size) {
                    smoothCurve?.set(i, base(smoothCurve!![i], MusicPlayer.getCurve()?.get(i) ?: return, 0.15).toFloat().toDouble())
                }
                val curve = MusicPlayer.getCurve() ?: return
                val fftSize = 1024
                val sampleRate = 44100.0
                val frequencies = DoubleArray(fftSize / 2) { i -> i * sampleRate / fftSize }

                val screenWidth = sr.scaledWidth.toFloat()
                val screenHeight = sr.scaledHeight.toFloat()

                for (i in curve.indices) {
                    val freq = frequencies[i]
                    val magnitude = curve[i].coerceIn(0.0, 1.0)

                    val x = (freq / 22050.0 * screenWidth).toFloat()
                    val height = (magnitude * 100f * MusicOverlay.amplitude.value.toFloat()).toFloat()

                    Render2DUtils.drawRect(
                        x,
                        screenHeight - height,
                        2f,
                        height,
                        MusicOverlay.color.rGB
                    )
                }

            }
        }
    }

    fun drawSong(x: Float, y: Float, width: Float, height: Float) {
        val sr = ScaledResolution(Utility.Companion.mc)
        val current = MusicPlayer.playList.getCurrent() as Music
        val s18 = FPSMaster.Companion.fontManager.s18
        Render2DUtils.drawOptimizedRoundedRect(x, y, width, height, Color(0, 0, 0, 180))
        Render2DUtils.drawOptimizedRoundedRect(x, y, songProgress, height, MusicOverlay.progressColor.color)
        songProgress = base(songProgress.toDouble(), (6 + (width - 6) * playProgress).toDouble(), 0.1).toFloat()
        Render2DUtils.drawImage(
            ResourceLocation("music/netease/" + current.id),
            x + 5,
            y + 5,
            height - 10,
            height - 10,
            -1
        )
        FPSMaster.Companion.fontManager.s18.drawString(
            current.name,
            x + 40,
            y + 6,
            FPSMaster.theme.textColorTitle.rgb
        )
        FPSMaster.Companion.fontManager.s16.drawString(
            current.author,
            x + 40,
            y + 18,
            FPSMaster.theme.textColorDescription.rgb
        )
    }
}
